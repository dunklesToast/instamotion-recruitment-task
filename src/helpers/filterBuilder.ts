import type { FormikFilters, HTTPFilterBody } from '@type/Filter.type';

/**
 * This function generates the filter for GQL. It's a bit ugly since I do not know how exactly the filtering
 * works and if I can send undefined values.
 *
 * @param filters
 */
export function buildFilters(filters: FormikFilters): HTTPFilterBody {
  const generatedFilters: HTTPFilterBody = {};

  if (filters.make) {
    generatedFilters['make-model'] = filters.make;
  }

  if (filters.model) {
    generatedFilters['make-model'] = `${filters.make}:${filters.model}`;
  }

  if (filters.category) {
    generatedFilters.category = filters.category;
  }

  if (filters.exteriorColor) {
    generatedFilters.exteriorColor = filters.exteriorColor;
  }

  if (filters.fuel) {
    generatedFilters.fuel = filters.fuel;
  }

  if (filters.gearbox) {
    generatedFilters.gearbox = filters.gearbox;
  }

  if (filters.mileage.from || filters.mileage.to) {
    generatedFilters.mileage = `["${filters.mileage.from}", "${filters.mileage.to}"]`;
  }

  if (filters.power.from || filters.power.to) {
    generatedFilters.power = `["${filters.power.from}", "${filters.power.to}"]`;
  }

  if (filters.price.from || filters.price.to) {
    generatedFilters.price = `["${filters.price.from}", "${filters.price.to}"]`;
  }

  return generatedFilters;
}
