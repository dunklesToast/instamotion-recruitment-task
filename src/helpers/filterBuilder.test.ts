import { buildFilters } from '@helpers/filterBuilder';
import type { FormikFilters } from '@type/Filter.type';

const basicFilter: FormikFilters = {
  mileage: {
    from: '',
    to: '',
  },
  power: {
    from: '',
    to: '',
  },
  price: {
    from: '',
    to: '',
  },
};

const fullFilter: FormikFilters = {
  category: 'AUTOMATIC_GEAR',
  exteriorColor: 'BLACK',
  fuel: 'ELECTRICITY',
  make: 'Opel',
  model: 'Adam',
  gearbox: 'AUTOMATIC',
  mileage: {
    from: '5',
    to: '10',
  },
  power: {
    from: '5',
    to: '10',
  },
  price: {
    from: '5',
    to: '10',
  },
};

test('FilterBuilder', () => {
  expect(buildFilters(basicFilter)).toMatchSnapshot();
  expect(buildFilters(fullFilter)).toMatchSnapshot();
});
