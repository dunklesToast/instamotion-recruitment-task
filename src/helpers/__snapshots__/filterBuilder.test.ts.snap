// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBuilder 1`] = `Object {}`;

exports[`FilterBuilder 2`] = `
Object {
  "category": "AUTOMATIC_GEAR",
  "exteriorColor": "BLACK",
  "fuel": "ELECTRICITY",
  "gearbox": "AUTOMATIC",
  "make-model": "Opel:Adam",
  "mileage": "[\\"5\\", \\"10\\"]",
  "power": "[\\"5\\", \\"10\\"]",
  "price": "[\\"5\\", \\"10\\"]",
}
`;
